### DISCUSSION TITLE
Why isn't a symlink converted to relative path?

### DISCUSSION THREAD

**Discussion by com-nvidia-yohuang** [ccache/ccache#553] [Questions & answers] [answered]
### Question ###
Hi Guys,

I'd like know why we can't convert the absolute file path to relative path if the file is symlink. When hashing the input file path, the abspath would cause cache miss, and actually we don't need to resolve the real path when hashing the input file I think.

thanks,
Justin

```
    if (Stat::lstat(argv[i], Stat::OnError::log).is_symlink()) {
      // Don't rewrite source file path if it's a symlink since
      // make_relative_path resolves symlinks using realpath(3) and this leads
      // to potentially choosing incorrect relative header files. See the
      // "symlink to source file" test.
      args_info.input_file = from_cstr(argv[i]);
    } else {
      // Rewrite to relative to increase hit rate.
      args_info.input_file = make_relative_path(ctx, argv[i]);
    }

```

**Comment by jrosdahl** [comment, marked as answer]
Hi,

> I'd like know why we can't convert the absolute file path to relative path if the [input] file is symlink.

The answer is in the comment you quoted:

```
// Don't rewrite source file path if it's a symlink since
// make_relative_path resolves symlinks using realpath(3) and this leads
// to potentially choosing incorrect relative header files. See the
// "symlink to source file" test.
```

> actually we don't need to resolve the real path when hashing the input file I think.

The test case mentioned in the comment proves why it's needed.

Or rather: why it _was_ needed. As it happens, I actually removed the use of realpath in e8444a3893cb just a week ago so the check in question can now be removed. Done in 6aba3fb3b48e.

**Comment by com-nvidia-yohuang** [comment]
Yes, you are right.

I debugged a similar problem to your test case long time ago but simply forgot it. 

thanks!
